easybuild/tools/containers.py

```diff
@@ -0,0 +1,304 @@
+                "--container-base docker:<image>:<tag>",
+            ])
+            raise EasyBuildError(error_msg)
+    else:
+        raise EasyBuildError("--container-base must be specified")
+
+    # first argument to --container-base is the Singularity bootstrap agent (localimage, shub, docker)
+    bootstrap_agent = base_specs[0]
+
+    # check bootstrap type value and ensure it is localimage, shub, docker
+    if bootstrap_agent not in SINGULARITY_BOOTSTRAP_TYPES:
+        known_bootstrap_agents = ', '.join(SINGULARITY_BOOTSTRAP_TYPES)
+        raise EasyBuildError("Bootstrap agent in container base spec must be one of: %s" % known_bootstrap_agents)
+
+    res = {'bootstrap_agent': bootstrap_agent}
+
+    for idx, base_spec in enumerate(base_specs[1:]):
+        res.update({'arg%d' % (idx + 1): base_spec})
+
+    return res
+
+
+def generate_singularity_recipe(easyconfigs, container_base):
+    """Main function to Singularity definition file and image."""
+
+    cont_path = container_path()
+
+    # check if --containerpath is valid path and a directory
+    if os.path.isdir(cont_path):
+        _log.info("Path for container recipes & images: %s", cont_path)
+    else:
+        raise EasyBuildError("Location for container recipes & images is a non-existing directory: %s" % cont_path)
+
+    base_specs = parse_container_base(container_base)
+
+    # extracting application name,version, version suffix, toolchain name, toolchain version from
+    # easyconfig class
+
+    bootstrap_agent = base_specs['bootstrap_agent']
+
+    base_image, base_image_tag = None, None
+
+    # with localimage it only takes 2 arguments. --container-base localimage:/path/to/image
+    # checking if path to image is valid and verify image extension is '.img' or '.simg'
+    if base_specs['bootstrap_agent'] == LOCALIMAGE:
+        base_image = base_specs['arg1']
+        if os.path.exists(base_image):
+            # get the extension of container image
+            image_ext = os.path.splitext(base_image)[1]
+            if image_ext == '.img' or image_ext == '.simg':
+                _log.debug("Extension for base container image to use is OK: %s", image_ext)
+            else:
+                raise EasyBuildError("Invalid image extension '%s' must be .img or .simg", image_ext)
+        else:
+            raise EasyBuildError("Singularity base image at specified path does not exist: %s", base_image)
+
+    # otherwise, bootstrap agent is 'docker' or 'shub'
+    # format --container-base {docker|shub}:<image>:<tag>
+    else:
+        base_image = base_specs['arg1']
+        # image tag is optional
+        base_image_tag = base_specs.get('arg2', None)
+
+    bootstrap_from = base_image
+    if base_image_tag:
+        bootstrap_from += ':' + base_image_tag
+
+    # if there is osdependencies in easyconfig then add them to Singularity recipe
+    install_os_deps = ''
+    osdeps = easyconfigs[0]['ec']['osdependencies']
```

Comment:
This is not enough as noted previously, I think you should not assume that only one build needs to have its os dependencies provided.

Reply:
I agree with @FooBarQuaxx easybuild will need to add all the osdeps from the dependency list ``--robot`` which could imply searching all easyconfigs. I'd recommend you create a set data structure and add os dependency and then iterate over the set and install the pkgs. 

When easyconfigs specify os dependency as ``osdependencies = [('libibverbs-dev', 'libibverbs-devel', 'rdma-core-devel')] `` then this could be processed as each element and added to the set. It would be okay to add libibverbs-dev and libibverbs-devel into the system since we know one of these packages are not going to be found. In the future it would be good to tag os dependency with the linux distribution so you can just pick the os dependency based on the bootstrap os in the container.

Reply:
For now, I'll change this to consider OS deps of all easyconfigs that are passed down to here.

I would like to leave also taking into account OS dependencies after resolving dependencies for future enhancement (you'll be able to get around this by specifying the easyconfig on the `eb` command for the time being).